### PR TITLE
Deflake C++ qps tests and improve verbosity of driver reporting

### PR DESCRIPTION
**Pull request by vjpai**
This fixes some races in the test that were causing problems at shutdown and reporting times.
This also includes #7244 to replace assertions in the driver with more useful error reporting.

### CODE REVIEW

**Comment by vjpai** [comment]
If this works, this fixes #6251, #6540, and #6844

**Comment by vjpai** [comment]
This is not quite ready but the failure is reproducible so it can be debugged easily.

File: test/cpp/qps/server_async.cc [hunk 1]
```diff
@@ -123,21 +123,22 @@ class AsyncQpsServerTest : public Server {
...
       threads_.emplace_back(&AsyncQpsServerTest::ThreadFunc, this, i);
     }
   }
   ~AsyncQpsServerTest() {
     for (auto ss = shutdown_state_.begin(); ss != shutdown_state_.end(); ++ss) {
-      (*ss)->set_shutdown();
+      std::lock_guard<std::mutex> lock((*ss)->mutex);
+      (*ss)->shutdown = true;
     }
     server_->Shutdown();
```
Comment (ctiller) [inline test/cpp/qps/server_async.cc]: I added a deadline here in #6737 to deflake (with good success)

**Comment by vjpai** [comment]
Performance smoke test passes, including some of the ones that had been failing.

**Comment by vjpai** [comment]
@ctiller if you're too busy, please feel free to reassign.

**Comment by ctiller** [comment]
Sent you #7407 which is this + merge with head, since I broke it.

**Pull request merged**